# **CSAI 472 - Speech Recognition**
# **Final Project Implementation**
# **Speech-to-Text, Translation, and Text-to-Speech**
## **Team 2:**
- Zeina Ayman - ID: 202200351
- Mai Waheed - ID: 202200556
- Lujain Ahmad - ID: 202201738

**Input Audio --> Speech-to-Text --> Translate English Text to Arabic / French / Spanish --> Translated Text-to-Speech**


# **1. Importing Libraries**


In [ ]:
!pip install -q openai-whisper deep-translator gTTS
!sudo apt-get install -y ffmpeg

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 803.2/803.2 kB 11.9 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.3/42.3 kB 3.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.2/98.2 kB 10.9 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
typer 0.25.1 requires click>=8.2.1, but you have click 8.1.8 which is incompatible.
wandb 0.27.0 requires click>=8.2.0, but you have click 8.1.8 which is incompatible.
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
ffmpeg is already the newest version (7:4.4.2-0ubuntu0.22.04.1).
0 upgraded, 0 newly installed, 0 to remove and 51 not upgraded.


In [ ]:
import whisper
from deep_translator import GoogleTranslator
from gtts import gTTS
from IPython.display import Audio, display
import urllib.request

**All packages and tools needed for the project (speech recognition, translation, and text-to-speech) were installed successully.**

# **2. Loading Input Audio Files**

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
short_audio_path = "/content/drive/MyDrive/Speech_Final_Project/English_Short_Input.mp4"
long_audio_path = "/content/drive/MyDrive/Speech_Final_Project/English_Long_Input.mp4"
noisy_audio_path = "/content/drive/MyDrive/Speech_Final_Project/English_Noisy_Input.mp4"

In [ ]:
print("English Short Audio:")
display(Audio(short_audio_path))
print("")
print("English Long Audio:")
display(Audio(long_audio_path))
print("")
print("English Noisy Audio:")
display(Audio(noisy_audio_path))

English Short Audio:



English Long Audio:



English Noisy Audio:


**3 english audio files were extracted from google drive.**

# **3. Speech-to-Text Conversion**

In [ ]:
model = whisper.load_model("small")

decode_opts = dict(
    temperature=0,
    beam_size=5,
    condition_on_previous_text=False,
)

result_1 = model.transcribe(short_audio_path, **decode_opts)
result_2 = model.transcribe(long_audio_path,  **decode_opts)
result_3 = model.transcribe(noisy_audio_path, **decode_opts)

transcribed_text_1 = result_1["text"]
transcribed_text_2 = result_2["text"]
transcribed_text_3 = result_3["text"]

100%|███████████████████████████████████████| 461M/461M [00:07<00:00, 67.7MiB/s]


In [ ]:
print("Short English Transcribed Text:")
print(transcribed_text_1)
print("")
print("Long English Transcribed Text:")
print(transcribed_text_2)
print("")
print("Noisy English Transcribed Text:")
print(transcribed_text_3)

Short English Transcribed Text:
 What's a lesson you wish you learned earlier in life?

Long English Transcribed Text:
 Raise your hand if you have recently had to introduce yourself or make small talk. Yes, that is what today is all about. How to speak better in the moment.

Noisy English Transcribed Text:
 It it's all connected if we go back and look at it and there's a whole lot of I thought that was the end Well, it was the end, but it was the beginning of this thing What you're leaning into is that mystery going forward, right?


**Whisper Small model was used to transcribe English audio files, including short, long, and noisy samples, yielding accurate text transcriptions for all.**

# **4. Language Translation**

In [ ]:
translated_arabic_text_1 = GoogleTranslator(source="auto", target="ar").translate(transcribed_text_1)
translated_arabic_text_2 = GoogleTranslator(source="auto", target="ar").translate(transcribed_text_2)
translated_arabic_text_3 = GoogleTranslator(source="auto", target="ar").translate(transcribed_text_3)

In [ ]:
print("Translated Short Arabic Text:")
print(translated_arabic_text_1)
print("")
print("Translated Long Arabic Text:")
print(translated_arabic_text_2)
print("")
print("Translated Noisy Arabic Text:")
print(translated_arabic_text_3)

Translated Short Arabic Text:
ما هو الدرس الذي تتمنى أن تتعلمه في وقت سابق من حياتك؟

Translated Long Arabic Text:
ارفع يدك إذا اضطررت مؤخرًا إلى تقديم نفسك أو إجراء محادثة قصيرة. نعم، هذا هو ما يدور حوله اليوم. كيف تتحدث بشكل أفضل في هذه اللحظة

Translated Noisy Arabic Text:
كل شيء متصل إذا عدنا ونظرنا إليه وهناك الكثير من الأشياء التي اعتقدت أنها كانت النهاية حسنًا، لقد كانت النهاية، لكنها كانت بداية هذا الشيء الذي تميل إليه هو المضي قدمًا في هذا الغموض، أليس كذلك؟


In [ ]:
translated_french_text_1 = GoogleTranslator(source="auto", target="fr").translate(transcribed_text_1)
translated_french_text_2 = GoogleTranslator(source="auto", target="fr").translate(transcribed_text_2)
translated_french_text_3 = GoogleTranslator(source="auto", target="fr").translate(transcribed_text_3)

In [ ]:
print("Translated Short French Text:")
print(translated_french_text_1)
print("")
print("Translated Long French Text:")
print(translated_french_text_2)
print("")
print("Translated Noisy French Text:")
print(translated_french_text_3)

Translated Short French Text:
Quelle est la leçon que vous auriez aimé apprendre plus tôt dans la vie ?

Translated Long French Text:
Levez la main si vous avez récemment dû vous présenter ou bavarder. Oui, c’est de cela qu’il s’agit aujourd’hui. Comment mieux parler sur le moment.

Translated Noisy French Text:
Tout est lié si nous y retournons et y regardons et il y a beaucoup de choses. Je pensais que c'était la fin. Eh bien, c'était la fin, mais c'était le début de cette chose. Ce sur quoi vous vous penchez, c'est ce mystère qui va de l'avant, n'est-ce pas ?


In [ ]:
translated_spanish_text_1 = GoogleTranslator(source="auto", target="es").translate(transcribed_text_1)
translated_spanish_text_2 = GoogleTranslator(source="auto", target="es").translate(transcribed_text_2)
translated_spanish_text_3 = GoogleTranslator(source="auto", target="es").translate(transcribed_text_3)

In [ ]:
print("Translated Short Spanish Text:")
print(translated_spanish_text_1)
print("")
print("Translated Long Spanish Text:")
print(translated_spanish_text_2)
print("")
print("Translated Noisy Spanish Text:")
print(translated_spanish_text_3)

Translated Short Spanish Text:
¿Cuál es una lección que desearías haber aprendido antes en la vida?

Translated Long Spanish Text:
Levante la mano si recientemente ha tenido que presentarse o entablar una pequeña charla. Sí, de eso se trata hoy. Cómo hablar mejor en el momento.

Translated Noisy Spanish Text:
Está todo conectado si retrocedemos y lo miramos y hay muchas cosas que pensé que era el final. Bueno, fue el final, pero fue el comienzo de esto. En lo que te estás inclinando es en ese misterio que sigue adelante, ¿verdad?


**Translated the transcribed English text from audio files into Arabic, French, and Spanish using Google Translator.**

# **5. Text-to-Speech Conversion**

In [ ]:
tts_ar_1 = gTTS(text=translated_arabic_text_1, lang="ar")
tts_ar_2 = gTTS(text=translated_arabic_text_2, lang="ar")
tts_ar_3 = gTTS(text=translated_arabic_text_3, lang="ar")

arabic_audio_path_1 = "arabic_translation_short.mp3"
tts_ar_1.save(arabic_audio_path_1)
arabic_audio_path_2 = "arabic_translation_long.mp3"
tts_ar_2.save(arabic_audio_path_2)
arabic_audio_path_3 = "arabic_translation_noisy.mp3"
tts_ar_3.save(arabic_audio_path_3)

In [ ]:
print("Arabic Short Audio:")
display(Audio(arabic_audio_path_1))
print("")
print("Arabic Long Audio:")
display(Audio(arabic_audio_path_2))
print("")
print("Arabic Noisy Audio:")
display(Audio(arabic_audio_path_3))

Arabic Short Audio:



Arabic Long Audio:



Arabic Noisy Audio:


In [ ]:
tts_fr_1 = gTTS(text=translated_french_text_1, lang="fr")
tts_fr_2 = gTTS(text=translated_french_text_2, lang="fr")
tts_fr_3 = gTTS(text=translated_french_text_3, lang="fr")

french_audio_path_1 = "french_translation_short.mp3"
tts_fr_1.save(french_audio_path_1)
french_audio_path_2 = "french_translation_long.mp3"
tts_fr_2.save(french_audio_path_2)
french_audio_path_3 = "french_translation_noisy.mp3"
tts_fr_3.save(french_audio_path_3)

In [ ]:
print("French Short Audio:")
display(Audio(french_audio_path_1))
print("")
print("French Long Audio:")
display(Audio(french_audio_path_2))
print("")
print("French Noisy Audio:")
display(Audio(french_audio_path_3))

French Short Audio:



French Long Audio:



French Noisy Audio:


In [ ]:
tts_es_1 = gTTS(text=translated_spanish_text_1, lang="es")
tts_es_2 = gTTS(text=translated_spanish_text_2, lang="es")
tts_es_3 = gTTS(text=translated_spanish_text_3, lang="es")

spanish_audio_path_1 = "spanish_translation_short.mp3"
tts_es_1.save(spanish_audio_path_1)
spanish_audio_path_2 = "spanish_translation_long.mp3"
tts_es_2.save(spanish_audio_path_2)
spanish_audio_path_3 = "spanish_translation_noisy.mp3"
tts_es_3.save(spanish_audio_path_3)

In [ ]:
print("Spanish Short Audio:")
display(Audio(spanish_audio_path_1))
print("")
print("Spanish Long Audio:")
display(Audio(spanish_audio_path_2))
print("")
print("Spanish Noisy Audio:")
display(Audio(spanish_audio_path_3))

Spanish Short Audio:



Spanish Long Audio:



Spanish Noisy Audio:


**gTTS was used to convert translated Arabic, French, and Spanish text into speech, successfully generating and playing audio files.**

# **6. Saving Generated Audio Files to Google Drive**

In [ ]:
drive_arabic_path_1 = "/content/drive/MyDrive/Speech_Final_Project/arabic_translation_short.mp3"
drive_arabic_path_2 = "/content/drive/MyDrive/Speech_Final_Project/arabic_translation_long.mp3"
drive_arabic_path_3 = "/content/drive/MyDrive/Speech_Final_Project/arabic_translation_noisy.mp3"

tts_ar_1.save(drive_arabic_path_1)
tts_ar_2.save(drive_arabic_path_2)
tts_ar_3.save(drive_arabic_path_3)

print("Arabic Audio Saved to Google Drive")

Arabic Audio Saved to Google Drive


In [ ]:
drive_french_path_1 = "/content/drive/MyDrive/Speech_Final_Project/french_translation_short.mp3"
drive_french_path_2 = "/content/drive/MyDrive/Speech_Final_Project/french_translation_long.mp3"
drive_french_path_3 = "/content/drive/MyDrive/Speech_Final_Project/french_translation_noisy.mp3"

tts_fr_1.save(drive_french_path_1)
tts_fr_2.save(drive_french_path_2)
tts_fr_3.save(drive_french_path_3)

print("French Audio Saved to Google Drive")

French Audio Saved to Google Drive


In [ ]:
drive_spanish_path_1 = "/content/drive/MyDrive/Speech_Final_Project/spanish_translation_short.mp3"
drive_spanish_path_2 = "/content/drive/MyDrive/Speech_Final_Project/spanish_translation_long.mp3"
drive_spanish_path_3 = "/content/drive/MyDrive/Speech_Final_Project/spanish_translation_noisy.mp3"

tts_es_1.save(drive_spanish_path_1)
tts_es_2.save(drive_spanish_path_2)
tts_es_3.save(drive_spanish_path_3)

print("Spanish Audio Saved to Google Drive")

Spanish Audio Saved to Google Drive


**Saved the generated Arabic, French, and Spanish speech files to Google Drive for permanent storage.**

# **7. System Latency and Real-Time Capability**

In [ ]:
import time

audio_path = noisy_audio_path
target_lang = "ar"

t0 = time.time()

# speech-to-text
text = model.transcribe(audio_path, **decode_opts)["text"]
t1 = time.time()

# translation
translated = GoogleTranslator(source="auto", target=target_lang).translate(text)
t2 = time.time()

# text-to-speech
gTTS(text=translated, lang=target_lang).save("timing_test.mp3")
t3 = time.time()

print("=== English to Arabic ===")
print(f"Speech-to-Text:  {t1 - t0:.2f} seconds")
print(f"Translation:     {t2 - t1:.2f} seconds")
print(f"Text-to-Speech:  {t3 - t2:.2f} seconds")
print(f"Total Pipeline Time:  {t3 - t0:.2f} seconds")

=== English to Arabic ===
Speech-to-Text:  1.70 seconds
Translation:     0.19 seconds
Text-to-Speech:  0.76 seconds
Total Pipeline Time:  2.66 seconds


In [ ]:
import time

audio_path = noisy_audio_path
target_lang = "fr"

t0 = time.time()

# speech-to-text
text = model.transcribe(audio_path, **decode_opts)["text"]
t1 = time.time()

# translation
translated = GoogleTranslator(source="auto", target=target_lang).translate(text)
t2 = time.time()

# text-to-speech
gTTS(text=translated, lang=target_lang).save("timing_test.mp3")
t3 = time.time()

print("=== English to French ===")
print(f"Speech-to-Text:  {t1 - t0:.2f} seconds")
print(f"Translation:     {t2 - t1:.2f} seconds")
print(f"Text-to-Speech:  {t3 - t2:.2f} seconds")
print(f"Total Pipeline Time:  {t3 - t0:.2f} seconds")

=== English to French ===
Speech-to-Text:  1.71 seconds
Translation:     0.21 seconds
Text-to-Speech:  1.61 seconds
Total Pipeline Time:  3.53 seconds


In [ ]:
import time

audio_path = noisy_audio_path
target_lang = "es"

t0 = time.time()

# speech-to-text
text = model.transcribe(audio_path, **decode_opts)["text"]
t1 = time.time()

# translation
translated = GoogleTranslator(source="auto", target=target_lang).translate(text)
t2 = time.time()

# text-to-speech
gTTS(text=translated, lang=target_lang).save("timing_test.mp3")
t3 = time.time()

print("=== English to Spanish ===")
print(f"Speech-to-Text:  {t1 - t0:.2f} seconds")
print(f"Translation:     {t2 - t1:.2f} seconds")
print(f"Text-to-Speech:  {t3 - t2:.2f} seconds")
print(f"Total Pipeline Time:  {t3 - t0:.2f} seconds")

=== English to Spanish ===
Speech-to-Text:  1.77 seconds
Translation:     0.20 seconds
Text-to-Speech:  1.26 seconds
Total Pipeline Time:  3.23 seconds


**Measured the execution time of each stage of the pipeline for Arabic, French, and Spanish, processed the noisy audio in 2.66–3.53 seconds. Speech-to-Text was the most time-consuming stage, while translation was the fastest.**